A43-sdk-authorization.md

```diff
@@ -0,0 +1,424 @@
+     any user (See example below). Similarly, if only request is empty, we evaluate
+     against the source fields and apply that to any action.
+   - This also applies to sub fields, like if the principals in source is empty, the
+     behavior would be similar to an empty source.
+2. Each source could contain a list of principals. The principals are ORed together,
+   i.e. it matches if one of them matches.
+   Sequence of steps to evaluate each principal from config -
+   1. If TLS is not used, matching fails.
+   2. A match is found if principal matches any one of URI SANs or DNS SANs or Subject
+      field from x509 client certificate.
+   3. We also get a match if there is no client certificate, and principal is an empty
+      string.
+3. Each request could contain a list of URL paths (i.e. fully qualified RPC methods)
+   and list of http headers to match. Refer JSON schema above to understand matching
+   semantics.
+
+#### Example
+
+In the following policy example
+- Peer identity from ["spiffe://foo.com/sa/admin1", "spiffe://foo.com/sa/admin2"] is 
+  authorized to access any RPC methods in pkg.service
+- Any authenticated user is allowed to access "foo" and "bar" RPC methods.
+- Nobody can access "secret" RPC method, not even the admins as deny rules are evaluated 
+  first.
+
+```json
+{
+	"name": "example-policy",
+	"allow_rules": [
+		{
+			"name": "admin-access",
+			"source": {
+				"principals": [
+					"spiffe://foo.com/sa/admin1",
+					"spiffe://foo.com/sa/admin2"
+				]
+			},
+			"request": {
+				"paths": ["/pkg.service/*"]
+			}
+		},
+		{
+			"name": "dev-access",
+			"source": {
+				"principals": [
+					"*"
+				]
+			},
+			"request": {
+				"paths": [
+					"/pkg.service/foo",
+					"/pkg.service/bar"
+				]
+			}
+		}
+	],
+	"deny_rules": [
+		{
+			"name": "deny-access",
+			"request": {
+				"paths": [
+					"*/secret"
+				]
+			}
+		}
+	]
+}
+```
+
+### API
+
+gRPC will support both static initialization and dynamically reloading the policy
+from filesystem. In static initialization, the policy will be provided as a JSON
+string. In dynamic file reloading, the application will specify the file path that
+contains the authorization policy in JSON format.
+
+We recommend the users to use a single SDK authorization policy per gRPC server.
+If there are multiple policies, then there is a possibility that all the policies
+may not be evaluated against. For ex. if we have two policies for two different
+services say service A and service B. RPC to service B may get rejected, without
+even evaluating against service B policy, because it is evaluated after service A
+policy. On getting no match, service A policy could deny by default.
+
+Following code snippets show how to enable authorization in gRPC servers in 
+different languages.
+
+#### C++
+
+1. Static Initialization
+
+```C++
+std::shared_ptr<AuthorizationPolicyProviderInterface> provider = 
+	StaticDataAuthorizationPolicyProvider::Create(authz_policy);
```

Comment:
What happens if the policy is invalid?

Reply:
Oh sorry! I forgot to add the updates to API when I copied this from design document. Updated the snippet to match implementation.
If the policy is invalid, it returns null provider, and we can view the error status in status parameter.